account details screen: Display deactivation information for user

agrawal-d opened the pull request:
Fixes #3551.

Two changes:
1. Deactivation status is shown.
1. Button text is "View private messages" instead of "Send private message"

**Before**: (Assume Cyphase is deactivated ) 
![before](https://user-images.githubusercontent.com/7714968/72449566-e6a40e00-37de-11ea-90cb-42d3e28cccf8.png)

**After**:
![after](https://user-images.githubusercontent.com/7714968/72502566-a3d64a80-385f-11ea-8955-a295267b1e31.png)

I had to fiddle around a  lot to understand to get the relevant data, here is the reasoning why it is done in this manner:
 - The get-all-users *is not* used by the mobile app to fetch user lists ( this is the list that contains is_deactivated data ). Instead, /register event queue is used.
- So, there is no direct way to get if a user is deactivated. However, `getActiveUsersByEmail()` returns only activated users, comparing it with current user can give us `is_deactivated`.
- A lot of informative and detailed comments are present in `/src/api/modelTypes.js:47` and  `/src/users/userSelectors.js`, the reviewer may go through them.

I *really* enjoyed doing this :D

agrawal-d commented:
On second thoughts, removing the PM button is a wrong idea, because one might want to view past messages. I will revert the PM button, and instead, create a separate PR for disabling the send button on deactivated profile PM, and also attempt to fix the @undefined bug.

agrawal-d commented:
In a separate Pull Request, I will fix the _PM deactivated person screen_, as there are three bugs when PMing a deactivated profile
![image](https://user-images.githubusercontent.com/7714968/72502825-41ca1500-3860-11ea-8cd0-b633daf952ae.png)

   1. The user shows up as @undefined in the compose box.
   2. Actually sending a message results in an indefinite spinner being shown.
   3. While the send button is disabled, it is still possible to send attachments/images ( which get infinite spinners, this is tracked by a different issue (https://github.com/zulip/zulip-mobile/issues/3784)

All three will be fixed by simply removing the compose box, and showing a view similar to the #announce one:
![image](https://user-images.githubusercontent.com/7714968/72503010-ba30d600-3860-11ea-821a-d85e66e36f48.png)

I have created an issue for this : https://github.com/zulip/zulip-mobile/issues/3812

agrawal-d commented:
@gnprice @agrawal-d  please review :)

jainkuniya commented on src/account-info/AccountDetailsScreen.js:
```diff
@@ -29,29 +36,51 @@ class AccountDetailsScreen extends PureComponent<Props> {
+    let privateMessageButton = null;
+    if (isDeactivated) {
+      deactivatedText = (
+        <RawLabel style={styles.deactivatedText} text="(This user has been deactivated)" />
```
Can we use `<Label />` instead so that this text gets translated in user selected language.

agrawal-d replied:
I had used label first. It does not adapt to theme changes ( dark mode ).

jainkuniya replied:
Can you please provide more details on this? maybe screenshot.

Then if there are some layout issues, then it will be better to first enhance `Label` to make it look better in dark theme.

agrawal-d replied:
Is there a way that text can be translated, before it is added to the RawLabel?

agrawal-d replied:
**Fixed**

jainkuniya commented on src/account-info/AccountDetailsScreen.js:
```diff
@@ -29,29 +36,51 @@ class AccountDetailsScreen extends PureComponent<Props> {
+      deactivatedText = (
+        <RawLabel style={styles.deactivatedText} text="(This user has been deactivated)" />
+      );
+      privateMessageButton = (
```
Here both the components are same, just the difference is in `text`. This will look much more cleaner this this
```
text={isDeactivated ? "View private messages" : "Send private message"}
```

jainkuniya commented on src/account-info/AccountDetailsScreen.js:
```diff
@@ -29,29 +36,51 @@ class AccountDetailsScreen extends PureComponent<Props> {
+      privateMessageButton = (
         <ZulipButton
           style={styles.pmButton}
           text="Send private message"
```
I will prefer `Open chat`. This will work in both the cases.

agrawal-d replied:
Changed to `Open chat`

jainkuniya commented on src/account-info/AccountDetailsScreen.js:
```diff
@@ -29,29 +36,51 @@ class AccountDetailsScreen extends PureComponent<Props> {
 
 export default connectFlowFixMe((state: GlobalState, props) => ({
   user: getAccountDetailsUserForEmail(state, props.navigation.state.params.email),
+  isDeactivated:
```
Can you extract this to a selector? or better move this logic to `getAccountDetailsUserForEmail `, which by name says account details for email (and deactivated information belongs to this).

agrawal-d replied:
Done

agrawal-d commented:
@jainkuniya this is how test/label looks in dark mode, which is why I used RawLabel:
![image](https://user-images.githubusercontent.com/7714968/72681175-faa18580-3ae6-11ea-8845-bbf19d3521ee.png)

agrawal-d commented:
@jainkuniya  Have pushed some changes, please review. I have separated the deactivation text into a variable, so that it can be translated, and RawLabel can be used.

ray-kraesig commented on src/account-info/AccountDetailsScreen.js:
```diff
@@ -37,13 +43,20 @@ class AccountDetailsScreen extends PureComponent<Props> {
         <ZulipButton
           style={styles.pmButton}
-          text="Send private message"
+          text="Open chat"
```
The webapp prefers "private messages", eschewing "chat" in this context (and almost all others); we should follow suit. Your original suggestion is probably a better option.

(Also, a new string will need to be added to `messages_en.json` regardless.)

ray-kraesig commented on src/users/userSelectors.js:
```diff
@@ -159,13 +159,14 @@ export const getActiveUsersByEmail: Selector<Map<string, UserOrBot>> = createSel
   state => getAllUsersByEmail(state),
-  (email, allUsersByEmail) => {
+  (state, email) => getActiveUsersByEmail(state).has(email),
+  (email, allUsersByEmail, is_active) => {
```
Style nit: prefer `camelCase` to `snake_case` for anything that isn't being serialized to the network.

ray-kraesig commented on src/account-info/AccountDetailsScreen.js:
```diff
@@ -37,13 +43,20 @@ class AccountDetailsScreen extends PureComponent<Props> {
+    let deactivatedText = null;
+    if (user.is_active === false) {
+      deactivatedText = (
+        <Label style={styles.deactivatedText} text="(This user/bot has been deactivated)" />
```
This string needs to go in `messages_en.json`.

(Also, let's stick with `(This user has been deactivated)`, to mirror the webapp.)

agrawal-d commented:
@ray-kraesig have pushed some changes.

agrawal-d commented on src/users/userSelectors.js:
```diff
@@ -156,6 +156,23 @@ export const getActiveUsersByEmail: Selector<Map<string, UserOrBot>> = createSel
+    return new Map(deactivatedUsersByEmail);
+  },
+);
+
```
Would a selector that only returns a `boolean` be acceptable? Even I thought that n^2 is too much, but I thought that a selector must return a subset of the list of users.

agrawal-d replied:
I think It should be acceptable, because there exists `getOwnEmail`.

agrawal-d commented:
Thanks for reviewing @ray-kraesig. Have pushed some changes. Please have a look.

agrawal-d commented:
@ray-kraesig thanks for the review. I have made the changes that you requested:

- wrote a test for the new selector and separated that into a new commit
- fixed the flow type for the selector
- changed the docstring comment for the selector to be more clear as you suggested
- changed commit message to show change is in harmony with the Zulip webapp.

Please review. Thanks!

ray-kraesig commented on src/users/__tests__/userSelectors-test.js:
```diff
@@ -124,3 +125,12 @@ describe('getUsersSansMe', () => {
+    const deactivatedUser = eg.makeUser();
+    const state = eg.reduxState({ users });
+    expect(getUserIsActive(state, deactivatedUser.email)).toBeFalse();
+  });
```
You also need to check that it returns `true` for a user who has *not* been deactivated! And, less obviously, for a bot (`realm.crossRealmBots`).

Although the current implementation doesn't rely on it, you should probably include `deactivatedUser` in the redux-state's `realm.nonActiveUsers`. Alternatively, you can just reuse the state setup from several of the tests above.

(Optional useful trick: you can set up a shared `state` in the `describe` block and use it in its sibling `test` blocks.)

agrawal-d commented:
Yikes :scream: thanks to your suggestion to add a test for `cross realm bots`, I quickly realized that these was a small bug in the selector I created - I used `getUsersByEmail` instead of `getActiveUsersByEmail`. `getUsersByEmail` does not contain cross reamp bots!

ray-kraesig commented:
Merged. (I've added `Fixes #3551.` to the latter commit, for ease of reference later, but the commits are otherwise unchanged.)

Thanks!
